Suspend processes aggressively on critical memory pressure

bnham opened the pull request:
#### 572ff90b0bbd4ec0a5be3e7e2c3664907de70d5f
<pre>
Suspend processes aggressively on critical memory pressure
<a href="https://bugs.webkit.org/show_bug.cgi?id=281702">https://bugs.webkit.org/show_bug.cgi?id=281702</a>
<a href="https://rdar.apple.com/problem/138150321">rdar://problem/138150321</a>

Reviewed by Chris Dumez.

When the system is under critical memory pressure, we should lower the suspension interval of
WebContent processes for some period of time to help alleviate pressure on the system. In this patch
we lower the suspension interval to 10 seconds for 30 minutes after critical memory pressure is
detected, although those values might have to be tuned. This is disabled by default and enabled by a
process pool configuration preference.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessActivityState.cpp:
(WebKit::webProcessSuspensionDelay):
(WebKit::WebProcessActivityState::updateWebProcessSuspensionDelay):
* Source/WebKit/UIProcess/WebProcessActivityState.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::memoryPressureStatusChanged):
(WebKit::WebProcessProxy::updateWebProcessSuspensionDelay):
(WebKit::operator&lt;&lt;):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::m_checkMemoryPressureStatusTimer):
(WebKit::criticalMemoryPressureCheckInterval):
(WebKit::WebProcessPool::checkMemoryPressureStatus):
(WebKit::WebProcessPool::defaultWebProcessSuspensionDelay):
(WebKit::WebProcessPool::webProcessSuspensionDelay const):
(WebKit::WebProcessPool::memoryPressureStatusChangedForProcess):
(WebKit::WebProcessPool::updateWebProcessSuspensionDelay):
(WebKit::WebProcessPool::updateWebProcessSuspensionDelayWithPacing):
* Source/WebKit/UIProcess/WebProcessPool.h:

When a WebProcess gets a critical memory pressure notification, forward it to the associated
WebProcessPool in UIProcess. The WebProcessPool will then ask all WebProcesses to suspend quickly
until 30 minutes have elapsed with the system in a non-critical pressure state.

We detect the pressure change in WebProcess rather than in UIProcess (e.g. via MemoryPressureHandler
in UIProcess) because the kernel paces memory pressure notifications. Due to the high jetsam
priority of the UIProcess, a long period can elapse between the system experiencing critical pressure
and the UIProcess learning about the issue.

* Source/WTF/wtf/WeakHashSet.h:

Add a cast to the templated type for takeAny. We already have a similar cast in the iterator. This
is necessary for WeakHashSet&lt;WebProcessProxy&gt;::takeAny to work, since WebProcessProxy defines a
custom WeakValueType.

* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::webProcessSuspensionDelay const): Deleted.
(API::PageConfiguration::setWebProcessSuspensionDelay): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _webProcessSuspensionDelay]): Deleted.
(-[WKWebViewConfiguration _setWebProcessSuspensionDelay:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Remove webProcessSuspensionDelay as a per-page preference as that was only used for debugging
purposes. We just read these values directly in WebProcessPool via CFPreferences instead. That is
sufficient for debugging purposes without exposing this API surface. This doesn&apos;t break any clients
since only Safari was using it behind a -respondsToSelector: check.

* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.mm:
(-[_WKProcessPoolConfiguration suspendsWebProcessesAggressivelyOnCriticalMemoryPressure]):
(-[_WKProcessPoolConfiguration setSuspendsWebProcessesAggressivelyOnCriticalMemoryPressure:]):

Allow clients to enable aggressive suspension on critical memory pressure via a process pool config
value, which defaults to false.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottlerTimedActivity::ProcessThrottlerTimedActivity):
(WebKit::ProcessThrottlerTimedActivity::activityTimedOut):
(WebKit::ProcessThrottlerTimedActivity::setTimeout):
* Source/WebKit/UIProcess/ProcessThrottler.h:

Allow clients to change the timeout for a timed activity after it is created.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateWebProcessSuspensionDelay):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):

Stop logging the diagnostic key that depended on WebProcessProxy::isUnderMemoryPressure as it&apos;s not
used anymore. If we want something like this, then we should log this using the memory limit logging
that we added in 275198@main (which is only enabled on Mac for now.)

* Source/WTF/wtf/MemoryPressureHandler.h:
(WTF::MemoryPressureHandler::memoryPressureStatus const):
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/285426@main">https://commits.webkit.org/285426@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df05a011bdf3db0b507620dc1f67a97d75203bb9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25396 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76786 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23811 "Hash df05a011 for PR 35389 does not build (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59828 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23632 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/23811 "Hash df05a011 for PR 35389 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75665 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47078 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62514 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/72105 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22161 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65731 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20340 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78457 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71854 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16844 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64851 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16032 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13145 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6795 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93633 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47821 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/20609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48888 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50183 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessActivityState.cpp:
```diff
@@ -172,6 +179,20 @@ bool WebProcessActivityState::hasValidOpeningAppLinkActivity() const
+
+void WebProcessActivityState::updateWebProcessSuspensionDelay()
+{
+    Seconds timeout = std::visit(WTF::makeVisitor([&](WeakRef<WebPageProxy> page) {
```
Why not `const WeakRef<WebPageProxy>&` for this one?

bnham replied:
Will fix

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2051,6 +2051,25 @@ void WebProcessProxy::logDiagnosticMessageForResourceLimitTermination(const Stri
+    m_memoryPressureStatus = status;
+
+#if ENABLE(WEB_PROCESS_SUSPENSION_DELAY)
+    protectedProcessPool()->memoryPressureStatusChangedForProcess(Ref { *this }, status);
```
`Ref { }` is unnecessary.

bnham replied:
Hm yeah, I guess the caller is supposed to already protect us on the stack, so I'll fix that.

However looking at this a bit more closely, I think I need to change this to something like:

```
if (RefPtr pool = m_processPool.get())
   pool->memoryPressureStatusChangedForProcess
```

Because in theory, while cached/prewarmed processes should be suspended, a low memory notification could sneak in just before they're suspended. If this function runs on a cached or suspended WebProcessProxy, then `protectedProcessPool()` crashes.

bnham committed:
Suspend processes aggressively on critical memory pressure
https://bugs.webkit.org/show_bug.cgi?id=281702
rdar://problem/138150321

Reviewed by Chris Dumez.

When the system is under critical memory pressure, we should lower the suspension interval of
WebContent processes for some period of time to help alleviate pressure on the system. In this patch
we lower the suspension interval to 10 seconds for 30 minutes after critical memory pressure is
detected, although those values might have to be tuned. This is disabled by default and enabled by a
process pool configuration preference.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessActivityState.cpp:
(WebKit::webProcessSuspensionDelay):
(WebKit::WebProcessActivityState::updateWebProcessSuspensionDelay):
* Source/WebKit/UIProcess/WebProcessActivityState.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::memoryPressureStatusChanged):
(WebKit::WebProcessProxy::updateWebProcessSuspensionDelay):
(WebKit::operator<<):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::m_checkMemoryPressureStatusTimer):
(WebKit::criticalMemoryPressureCheckInterval):
(WebKit::WebProcessPool::checkMemoryPressureStatus):
(WebKit::WebProcessPool::defaultWebProcessSuspensionDelay):
(WebKit::WebProcessPool::webProcessSuspensionDelay const):
(WebKit::WebProcessPool::memoryPressureStatusChangedForProcess):
(WebKit::WebProcessPool::updateWebProcessSuspensionDelay):
(WebKit::WebProcessPool::updateWebProcessSuspensionDelayWithPacing):
* Source/WebKit/UIProcess/WebProcessPool.h:

When a WebProcess gets a critical memory pressure notification, forward it to the associated
WebProcessPool in UIProcess. The WebProcessPool will then ask all WebProcesses to suspend quickly
until 30 minutes have elapsed with the system in a non-critical pressure state.

We detect the pressure change in WebProcess rather than in UIProcess (e.g. via MemoryPressureHandler
in UIProcess) because the kernel paces memory pressure notifications. Due to the high jetsam
priority of the UIProcess, a long period can elapse between the system experiencing critical pressure
and the UIProcess learning about the issue.

* Source/WTF/wtf/WeakHashSet.h:

Add a cast to the templated type for takeAny. We already have a similar cast in the iterator. This
is necessary for WeakHashSet<WebProcessProxy>::takeAny to work, since WebProcessProxy defines a
custom WeakValueType.

* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::webProcessSuspensionDelay const): Deleted.
(API::PageConfiguration::setWebProcessSuspensionDelay): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _webProcessSuspensionDelay]): Deleted.
(-[WKWebViewConfiguration _setWebProcessSuspensionDelay:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Remove webProcessSuspensionDelay as a per-page preference as that was only used for debugging
purposes. We just read these values directly in WebProcessPool via CFPreferences instead. That is
sufficient for debugging purposes without exposing this API surface. This doesn't break any clients
since only Safari was using it behind a -respondsToSelector: check.

* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.mm:
(-[_WKProcessPoolConfiguration suspendsWebProcessesAggressivelyOnCriticalMemoryPressure]):
(-[_WKProcessPoolConfiguration setSuspendsWebProcessesAggressivelyOnCriticalMemoryPressure:]):

Allow clients to enable aggressive suspension on critical memory pressure via a process pool config
value, which defaults to false.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottlerTimedActivity::ProcessThrottlerTimedActivity):
(WebKit::ProcessThrottlerTimedActivity::activityTimedOut):
(WebKit::ProcessThrottlerTimedActivity::setTimeout):
* Source/WebKit/UIProcess/ProcessThrottler.h:

Allow clients to change the timeout for a timed activity after it is created.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateWebProcessSuspensionDelay):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):

Stop logging the diagnostic key that depended on WebProcessProxy::isUnderMemoryPressure as it's not
used anymore. If we want something like this, then we should log this using the memory limit logging
that we added in 275198@main (which is only enabled on Mac for now.)

* Source/WTF/wtf/MemoryPressureHandler.h:
(WTF::MemoryPressureHandler::memoryPressureStatus const):
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/285426@main
